File: lib/accounting/categories.ts
import categoryData from "@/workers/accounting_worker/engine/categories.json" with { type: "json" };

/**
 * The accounting category vocabulary — the TypeScript view of the one canonical
 * file the Python worker reads.
 *
 * Four vocabularies had grown apart: what the worker stored (29 strings), what
 * the professional chart and the AI produced, what this dropdown offered (27),
 * and what AI validation accepted (33) — 46 distinct strings in total. Eleven
 * categories the worker wrote could not be selected here at all, nine options
 * here were produced by nothing, and the same account appeared under several
 * spellings.
 *
 * A reviewer cannot correct a category the dropdown does not contain, and a
 * correction made in one spelling trains a learned rule the other side does not
 * recognise. So both languages now read the same file. It lives in the worker
 * package because the worker deploys from there and must have it at runtime.
 */

type CategoryEntry = { id: string; label: string; aliases: string[] };

const ENTRIES: CategoryEntry[] = (categoryData.categories as CategoryEntry[]).map((entry) => ({
  id: entry.id,
  label: entry.label,
  aliases: entry.aliases ?? [],
}));

/** Every category that may be stored, in presentation order. */
export const CANONICAL_CATEGORIES: string[] = ENTRIES.map((entry) => entry.id);

/** What a reviewer picks from: the stored value plus the label they read. */
export const CATEGORY_OPTIONS: Array<{ value: string; label: string }> = ENTRIES.map((entry) => ({
  value: entry.id,
  label: entry.label,
}));

const ALIAS_INDEX = new Map<string, string>();
for (const entry of ENTRIES) {
  ALIAS_INDEX.set(normalise(entry.id), entry.id);
  for (const alias of entry.aliases) ALIAS_INDEX.set(normalise(alias), entry.id);
}

function normalise(value: string): string {
  return value.trim().replace(/\s+/g, " ").toLowerCase();
}

/**
 * Map any known spelling to its canonical id, or null if unrecognised.
 *
 * Null rather than a guess: an unrecognised category is worth surfacing, and
 * silently coercing it to Uncategorised would hide a real inconsistency.
 */
export function canonicaliseCategory(value: string | null | undefined): string | null {
  if (!value) return null;
  return ALIAS_INDEX.get(normalise(value)) ?? null;
}

export function isKnownCategory(value: string | null | undefined): boolean {
  return canonicaliseCategory(value) !== null;
}

/**
 * What to show for a stored value — including one written before unification.
 * An unrecognised value is shown as-is rather than hidden: the reviewer should
 * see what the row actually holds.
 */
/**
 * The options to offer for a row currently holding `value`.
 *
 * A row written before unification may hold a historical spelling. It is
 * prepended rather than dropped, so the dropdown always shows what the row
 * actually holds — a `<select>` whose value is absent from its options renders
 * blank, which would misrepresent the row as unclassified and invite a
 * correction the reviewer never meant to make.
 */
export function categoryOptionsFor(value: string | null | undefined): Array<{ value: string; label: string }> {
  if (!value || CANONICAL_CATEGORIES.includes(value)) return CATEGORY_OPTIONS;
  const canonical = canonicaliseCategory(value);
  const label = canonical
    ? `${categoryLabel(value)} (was: ${value})`
    : `${value} (unrecognised)`;
  return [{ value, label }, ...CATEGORY_OPTIONS];
}

export function categoryLabel(value: string | null | undefined): string {
  if (!value) return "Uncategorised";
  const canonical = canonicaliseCategory(value);
  if (!canonical) return value;
  return ENTRIES.find((entry) => entry.id === canonical)?.label ?? canonical;
}

File: workers/accounting_worker/engine/categories.py
"""The accounting category vocabulary, loaded from the one canonical file.

Four different vocabularies had grown apart: what classify_transaction stored
(29 strings), what professional_account and the AI produced, what the review
dropdown offered (27), and what the AI validator accepted (33). Their union was
46 distinct strings. Eleven categories the worker wrote could not be selected by
a reviewer at all, nine dropdown options were produced by nothing, and the same
account appeared under several spellings — "Insurance" and "Insurance Expense",
"SARS / Tax Suspense" and "Tax / SARS Suspense", "Courier / Freight" and
"Courier / Delivery".

That is not a cosmetic problem. A reviewer cannot correct a category the
dropdown does not contain, and a correction made in one spelling trains a
learned rule the other side does not recognise.

categories.json is the single source. TypeScript imports the same file, so the
two languages cannot drift; a test in each language pins the parity.

Ids are the STORED values, and were chosen to be the professional-chart strings
wherever one existed, so the majority of existing rows already match. Aliases
are historical spellings: accepted on read and rendered correctly, never written.
"""

from __future__ import annotations

import json
from functools import lru_cache
from pathlib import Path

_CATEGORY_FILE = Path(__file__).with_name("categories.json")


@lru_cache(maxsize=1)
def _load() -> dict:
    return json.loads(_CATEGORY_FILE.read_text())


@lru_cache(maxsize=1)
def canonical_categories() -> tuple[str, ...]:
    """Every category that may be stored, in presentation order."""
    return tuple(entry["id"] for entry in _load()["categories"])


@lru_cache(maxsize=1)
def category_labels() -> dict[str, str]:
    return {entry["id"]: entry["label"] for entry in _load()["categories"]}


@lru_cache(maxsize=1)
def _alias_index() -> dict[str, str]:
    index: dict[str, str] = {}
    for entry in _load()["categories"]:
        index[entry["id"].casefold()] = entry["id"]
        for alias in entry.get("aliases", ()):
            index[alias.casefold()] = entry["id"]
    return index


def canonicalise_category(value: str | None) -> str | None:
    """Map any known spelling to its canonical id, or None if unrecognised.

    Case- and whitespace-insensitive, because the difference between
    "Bank Charges" and "bank charges " is a typo, not a different account.
    Returns None rather than a guess: an unrecognised category is a fact worth
    surfacing, and silently coercing it to Uncategorised would hide it.
    """
    if not value:
        return None
    return _alias_index().get(" ".join(str(value).split()).casefold())


def is_canonical_category(value: str | None) -> bool:
    """True only for a value already in canonical form — aliases are not."""
    return bool(value) and value in set(canonical_categories())


def is_known_category(value: str | None) -> bool:
    """True for a canonical id or any historical alias of one."""
    return canonicalise_category(value) is not None


@lru_cache(maxsize=1)
def unresolved_categories() -> frozenset[str]:
    """Categories that mean "nothing could determine this", not a decision.

    Flagged explicitly in categories.json rather than matched by pattern. A
    regex over "review|suspense|uncategori" also catches
    "Meals / Groceries - Non Deductible Review", "SARS / Tax Suspense" and
    "Refund / Suspense" — all real decisions a reviewer can act on — so a
    pattern would quietly demote three genuine classifications.
    """
    return frozenset(
        category["id"] for category in _load()["categories"] if category.get("unresolved")
    )


def is_unresolved_category(value: str | None) -> bool:
    """Whether a category is a parking bucket rather than an accounting answer.

    Canonicalises first, so an alias of a parking bucket is recognised as one.
    """
    canonical = canonicalise_category(value)
    return canonical is not None and canonical in unresolved_categories()
